README.md
# multishot

> A lib for generating multiple screenshots

## Getting Started

### Dependencies

Currently only works on osx.

```shell
brew install webkit2png
```

### Install

```shell
npm install multishot --save-dev
```

```shell
npm install multishot -g
```

## The CLI

### Overview

```
Usage: multishot [opts]

Options:
  -o     Output directory                                        [default: process.cwd()]
  -t     Temp directory. Note: This directory will be truncated. [default: "/tmp/multishot/"]
  -u     List of urls to use.
  -f     File containing list of urls (One url per line).
  -d     Device Profile. Pass with no option to see available presets.
  -w     Sets viewport width.
  -h     Sets viewport height.
  -a     Sets user agent.
  --tpl  Template file.
```

### Usage Examples

```
multishot -f example/urls.txt -o ~/Desktop
```

```
multishot -u http://google.com -o ~/Desktop
```

#### Text file input

The text file must contain one url per line. Blank lines are permitted for formatting. You may optionally group urls by 
putting a group name on its own line. Note: group names cannot contain http.

See example/urls.txt for an example

## The API

### Usage

```js
var Screenshot = require('multishot').Screenshot;

var urls = [
    {
      group: 'Circles',
      url: 'http://leanin.org/circles/'
    },
    {
      group: 'Circles',
      url: 'http://leanin.org/circle-login/'
    },
    {
      group: 'Circles',
      url: 'http://leanin.org/circle-faqs/'
    },
    'http://leanin.org',
    {
      url: 'http://leanin.org/team/'
    },
    {
      group: 'News+Inspiration',
      url: 'http://leanin.org/news-inspiration/'
    },
  ];

var options = {
  temp: '/tmp/multishot/',
  output: '/tmp/multishot/',
  style: {
    bodyBackground: 'pink',
    custom: '.url { text-decoration: underline; }'
  }
};

var shot = new Screenshot(urls, options);

shot.on('progress', function() {
  // Update progress bar if you have one
});

shot.on('complete', function(file) {
  // do something on complete
});

shot.on('error', function(err) {
  // Something bad happened
});
```


### Events

`progress`  
Triggered every time an image is rendered and once on complete.

`complete`  
Triggered once the file image has been generated. Returns the path to composed image.

`error`  
Returns an exception.


### Options

`output`  
A path to render the final composed image. Will be created if it doesn't exist.

`temp`  
Temp path where screenshots will be taken. Will be created and truncated each run so be careful when setting. Default: /tmp/multishot/

`template`  
A handlebars HTML file to use as the combined template. See custom template section for more information.

## Custom Templates

Multishot uses a handlebars template to generate the combined image. While the default template is nice, you may want to customize it:

Example:
```javascript
var shot = new Screenshot(urls, {
  template: '/Users/WalterWhite/templates/multishot.html'
});
```

The bare minimum your template should have is:

```html
<!DOCTYPE html>
<html>
<head>
  <base href="{{base}}">
</head> 
<body>
  {{#each grouped}}
  <div class="group">
    <h2>{{@key}}</h2>
    {{#each this}}
    <div class="screenshot-container">
      <div class="url">{{this.url}}</div>
      <img src="{{this.group}}/{{this.index}}-full.png" class="screenshot">
    </div>
    {{/each}}
   </div>
  {{/each}}
</body>
</html>
```

Templates are passed the following variables:

`base`  
Base path to load images

`grouped`  
Object containing all grouped images. See example HTML above for usage.

`styles`  
Various different values that a template can use to customize the css. See Styles section for available values.

## Custom Styles

These are the default styles:

| Variable            | Value              |
| ------------------- | ------------------ |
| background          | #fff               |
| bodyFontFamily      | "Lato", sans-serif |
| bodyFontWeight      | 400                |
| bodyFontColor       | #7f8c8d            |
| bodyFontSize        | 12px               |
| bodyLineHeight      | 17px               |
| groupMargin         | 40px 0             |
| screenshotImgBorder | 4px solid #d35400  |
| h2FontSize          | 24px               |
| h2FontColor         | #2C3E50            |
| h2FontWeight        | 700                |
| h2Margin            | 2px 0              |
| urlFontStyle        | italic             |

You can override or add new variables by passing them as properties to the `styles` option value.

If you want to pass in a full block of css:

```javascript
var options = {
  styles: {
    custom: '.someSelector{ background: white }'
  }
};
```

The `custom` property can contain any css as long as its valid. If you need any more control you should look into writing a unique template to match your needs.


History.md

0.2.0 / 2013-10-06 
==================

 * Merge branch 'feature/passable-styles'
 * Merge pull request #25 from firstandthird/feature/webkit2png-options
 * Merge pull request #27 from firstandthird/bug/missing-repository
 * Updated package.json to have repository field.
 * Merge branch 'master' into feature/passable-styles
 * Renamed lib to match upcoming changes in another branch.
 * Merge branch 'master' into feature/webkit2png-options
 * Working profiles and mobile view. See readme for usage.
 * WIP - More options for webkit2png.
 * Fixed incorrect base path.
 * Release 0.1.3
 * Fixed group not getting a single url.
 * [#5] - Added a check is webkit2png is available. We might be able to use this to detect other screenshot apps down the road.
 * [#12] - Output directory defaults to current directory.
 * WIP - #18, #19 - User agent and widths
 * Addded a way to pass in styles to override defaults

0.1.4 / 2013-10-03 
==================

  * Fixed incorrect base path.

0.1.3 / 2013-10-03 
==================

  * Merge pull request #17 from firstandthird/feature/check-webkit2png
  * Merge pull request #16 from firstandthird/feature/default-output-directory
  * Merge pull request #22 from firstandthird/bug/grouped-doesnt-contain-single-url
  * Fixed group not getting a single url.
  * [#5] - Added a check is webkit2png is available. We might be able to use this to detect other screenshot apps down the road.
  * [#12] - Output directory defaults to current directory.


example/test.js
(function(){
  "use strict";

  var Screenshot = require('../').Screenshot;

  var urls = [
    {
      group: 'Circles',
      url: 'http://leanin.org/circles/'
    },
    {
      group: 'Circles',
      url: 'http://leanin.org/circle-login/'
    },
    {
      group: 'Circles',
      url: 'http://leanin.org/circle-faqs/'
    },
    'http://leanin.org',
    {
      url: 'http://leanin.org/team/'
    },

    {
      group: 'News+Inspiration',
      url: 'http://leanin.org/news-inspiration/'
    },
  ];

  var shot = new Screenshot(urls);

  shot.start();
})();

lib/presets.js
(function(){
  "use strict";

  var presets = {
    "iphone": {
      width: 320,
      agent: "Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5"
    },
    "iphone-landscape": {
      width: 480,
      agent: "Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5"
    },
    "iphone5-landscape": {
      width: 568,
      agent: "Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5"
    },
    "ipad": {
      width: 768,
      agent: "Mozilla/5.0 (iPad; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5"
    },
    "ipad-landscape": {
      width: 1024,
      agent: "Mozilla/5.0 (iPad; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5"
    }
  };

  module.exports = presets;
})();

lib/screenshot-html.js
/**
 * screenshot-service
 *
 * Captures screenshots for the specified urls.
 *
 * Usage:
 *  var url = 'http://example.com';
 *  var screenshot = new Screenshot(url);
 *
 *  var urls = ['http://example.com', 'http://example.com/about'];
 *  var screenshots = new Screenshot(urls);
 */

 (function(module){
  'use strict';

  var spawn = require('child_process').spawn;
  var fs = require('fs');
  var async = require('async');
  var glob = require("glob");
  var path = require('path');
  var mkdirp = require('mkdirp');
  var handlebars = require('handlebars');

  require('shelljs/global');

  if (!which('webkit2png')) {
    echo('This script requires webkit2png. http://www.paulhammond.org/webkit2png/');
    exit(1);
  }

  var childArgs = function(url, outputTmp, index) {
    return [
      '--fullsize',
      '--dir=' + outputTmp,
      '--filename=' + index,
      url
    ];
  };

  var Screenshot = function(url, options) {
    var self = this;

    this.options = options || {
      output: process.cwd(),
      temp: '/tmp/multishot',
    };

    this.options.template = this.options.template || path.resolve(__dirname, '../template/template.html');

    this.groups = ['Default'];
    this.grouped = {};

    this.template = fs.readFileSync(this.options.template, 'utf8');

    async.parallel([
      function(done) {
        mkdirp(self.options.output, done);
      },
      function(done) {
        mkdirp(self.options.temp, done);
      },
      function(done) {
        glob(self.options.temp + '/**/*.png', function(err, files) {
          if(err) {
            self.emit('error');
            done();
            return;
          }

          async.each(files, function(file, finish) {
            fs.unlink(file, finish);
          }, function(err) {
            if(err) {
              self.emit('error', err);
            }
            done();
          });
        });
      }
    ],
    function(){
      this.urls = (typeof url === 'string') ? [url] : url;

      this.validateGroups();

      // Process urls
      async.each(this.urls, this.processUrl.bind(this), this.screenshotsComplete.bind(this));
    }.bind(this));
  };

  Screenshot.prototype = Object.create(require('events').EventEmitter.prototype);

  Screenshot.prototype.validateGroups = function() {
    var i, c, url;

    for(i = 0, c = this.urls.length; i < c; i++) {
      url = this.urls[i];
      if((!!url) && (url.constructor === Object)) {
        if(!url.url) {
          // Falls back to nothing. Will be skipped.
          url.url = '';
        }

        if(!url.group) {
          url.group = 'Default';
        }
      } else {
        url = {
          group: 'Default',
          url: url
        };
      }

      url.index = i;

      this.urls[i] = url;

      if(this.groups.indexOf(url.group) === -1) {
        this.groups.push(url.group);
      }

      if(!this.grouped[url.group]) {
        this.grouped[url.group] = [];
      }

      this.grouped[url.group].push(url);
    }
  };

  Screenshot.prototype.processUrl = function(url, done) {
    // Check if item is empty (either empty param or blank line)
    if(!url) {
      this.emit('progress');
      done();
      return;
    }

    var parser = spawn('webkit2png', childArgs(url.url, path.normalize(this.options.temp + '/' + url.group + '/'), url.index));
    parser.on('exit', function(code) {
      this.emit('progress');
      done();
    }.bind(this));
  };

  Screenshot.prototype.screenshotsComplete = function() {
    var self = this;
    
    var data = {
      base: this.options.temp,
      grouped: this.grouped
    };

    var template = handlebars.compile(this.template);
    var html = template(data);
    var templateFile = path.normalize(this.options.temp + '/template.html');

    var date = new Date();
    var outputFile = "" + date.getFullYear() + (date.getMonth()+1) + date.getDate();

    fs.writeFileSync(templateFile, html, 'utf8');

    var parser = spawn('webkit2png', childArgs(templateFile, this.options.output, outputFile));
    parser.on('exit', function(code) {
      self.emit('complete', path.normalize(self.options.output + '/' + outputFile + '-full.png'));
    });
  };

  module.exports = Screenshot;
 })(module);

lib/screenshot-service.js
/**
 * screenshot-service
 *
 * Captures screenshots for the specified urls.
 *
 * Usage:
 *  var url = 'http://example.com';
 *  var screenshot = new Screenshot(url);
 *
 *  var urls = ['http://example.com', 'http://example.com/about'];
 *  var screenshots = new Screenshot(urls);
 */

 (function(module){
  'use strict';

  var spawn = require('child_process').spawn;
  var fs = require('fs');
  var async = require('async');
  var im = require('imagick');
  var glob = require("glob");
  var path = require('path');
  var mkdirp = require('mkdirp');

  var childArgs = function(url, outputTmp, index) {
    return [
      '--fullsize',
      '--dir=' + outputTmp,
      '--filename=' + index,
      url
    ];
  };

  var Screenshot = function(url, options) {
    var self = this;

    this.options = options || {
      output: '/tmp/multishot',
      temp: '/tmp/multishot'
    };

    this.options.backgroundColor = this.options.backgroundColor || '#ecf0f1';
    this.options.labelBackgroundColor = this.options.labelBackgroundColor || '#ffffff';
    this.options.labelTextColor = this.options.labelTextColor || '#232323';
    this.options.labelFont = this.options.labelFont || 'Helvetica-Bold';
    this.options.labelFontSize = this.options.labelFontSize || 18;

    this.groups = ['Default'];
    this.grouped = {};

    async.parallel([
      function(done) {
        mkdirp(self.options.output, done);
      },
      function(done) {
        mkdirp(self.options.temp, done);
      },
      function(done) {
        glob(self.options.temp + '/**/*.png', function(err, files) {
          if(err) {
            self.emit('err');
            done();
            return;
          }

          async.each(files, function(file, finish) {
            fs.unlink(file, finish);
          }, function(err) {
            if(err) {
              self.emit('error', err);
            }
            done();
          });
        });
      }
    ],
    function(){
      this.urls = (typeof url === 'string') ? [url] : url;

      this.validateGroups();

      // Process urls
      async.each(this.urls, this.processUrl.bind(this), this.screenshotsComplete.bind(this));
    }.bind(this));
  };

  Screenshot.prototype = Object.create(require('events').EventEmitter.prototype);

  Screenshot.prototype.validateGroups = function() {
    var i, c, url;

    for(i = 0, c = this.urls.length; i < c; i++) {
      url = this.urls[i];
      if((!!url) && (url.constructor === Object)) {
        if(!url.url) {
          // Falls back to nothing. Will be skipped.
          url.url = '';
        }

        if(!url.group) {
          url.group = 'Default';
        }
      } else {
        url = {
          group: 'Default',
          url: url
        };
      }

      url.index = i;

      this.urls[i] = url;

      if(this.groups.indexOf(url.group) === -1) {
        this.groups.push(url.group);
      }

      if(!this.grouped[url.group]) {
        this.grouped[url.group] = [];
      }

      this.grouped[url.group].push(url);
    }
  };

  Screenshot.prototype.processUrl = function(url, done) {
    // Check if item is empty (either empty param or blank line)
    if(!url) {
      this.emit('progress');
      done();
      return;
    }

    var parser = spawn('webkit2png', childArgs(url.url, path.normalize(this.options.temp + '/' + url.group + '/'), url.index));
    parser.on('exit', function(code) {
      this.emit('progress');
      done();
    }.bind(this));
  };

  Screenshot.prototype.screenshotsComplete = function() {
    var self = this;

    async.each(this.groups, function(item, done) {
      async.each(self.grouped[item], self.generateTile.bind(self), function() {
        this.composeImages(item, done);
      }.bind(self));
    }, function(){
      self.groupCombined();
    });
  };

  Screenshot.prototype.generateTile = function(url, complete) {
    var self = this;
    var textfile = path.normalize(this.options.temp + '/' + url.group + '/' + url.index + '-label.png');
    var file = path.normalize(this.options.temp + '/' + url.group + '/' + url.index + '-full.png');

    // Order matters!
    async.series([function(done){
      // Add a frame first
      console.log('Generating frame for:', url.url);
      im.convert([file , '-matte', '-bordercolor', self.options.backgroundColor, '-border', '10x0', '-background', self.options.backgroundColor, file], function(err) {
        if(err) {
          console.log(err);
          self.emit('error', err);
          return;
        }

        done();
      });
    }, function(done) {
      // Create the text label. Note: easier to do it in a seperate image. You have more control over placement.
      console.log('Generating label:', url.url);
      im.convert(['-bordercolor', self.options.labelBackgroundColor, '-border', '10', '-background', self.options.labelBackgroundColor, '-fill', self.options.labelTextColor, '-font', self.options.labelFont, '-pointsize', self.options.labelFontSize, 'label:' + url.url, textfile], function(err, meta) {
        if(err) {
          console.log(err);
          self.emit('error', err);
          return;
        }

        done();
      });
    }, function(done) {
      //More spacing on label
      console.log('Adding spacing to label:', url.url);
      im.convert([textfile, '-bordercolor', self.options.backgroundColor, '-border', '10', '-background', self.options.backgroundColor, textfile], function(err, meta) {
        if(err) {
          console.log(err);
          self.emit('error', err);
          return;
        }

        done();
      });
    }, function(done) {
      //chopping label
      console.log('Chopping off bottom padding:', url.url);
      im.convert([textfile, '-crop', '+0-10', '+repage', textfile], function(err, meta) {
        if(err) {
          console.log(err);
          self.emit('error', err);
          return;
        }

        done();
      });
    }, function(done) {
      // Place label on screenshot
      console.log('Placing label:', url.url);
      im.convert(['-background', self.options.backgroundColor, '-append', textfile, file, file], function(err) {
        if(err) {
          console.log(err);
          self.emit('error', err);
          return;
        }

        done();
      });
    }], function(){
      console.log('Finished labeling screenshot:', url.url);
      complete();
    });
  };

  Screenshot.prototype.composeImages = function(group, complete) {
    var date = new Date();
    var urls = this.grouped[group];
    var i, c;

    var args = [];
    args.push('+append');

    for(i = 0, c = urls.length; i < c; i++) {
      args.push(path.normalize(this.options.temp + '/' + urls[0].group + '/' + urls[i].index + '-full.png'));
    }

    var outputFile = path.normalize(this.options.output + '/' + urls[0].group + '/' + date.getFullYear() + (date.getMonth()+1) + date.getDate() + "-combined.png");

    args.push(outputFile);

    im.convert(args, function(err, meta){
      if (err) {
        console.log(err);
        this.emit('error', err);
        return;
      }

      this.emit('progress');
      complete();
    }.bind(this));
  };

  Screenshot.prototype.groupCombined = function() {
    var self = this;

    glob(path.normalize(this.options.output + '/') + '**/*-combined.png', function(err, files) {
      var i, c;
      var date = new Date();
      var args = ["-append"];
      var outputFile = path.normalize(self.options.output + '/' + date.getFullYear() + (date.getMonth()+1) + date.getDate() + "-grouped.png");

      if(err) {
        console.error('Error:', err);
        self.emit('error', err);
        return;
      }

      console.log('Grouping combined images...');

      for(i = 0, c = files.length; i < c; i++) {
        args.push(files[i]);
      }

      args.push(outputFile);

      im.convert(args, function(err, meta){
        if (err) {
          this.emit('error', err);
          return;
        }

        this.emit('progress');

        this.emit('complete', outputFile);
      }.bind(self));
    });
  };

  module.exports = Screenshot;
 })(module);

lib/screenshot.js
/**
 * screenshot-service
 *
 * Captures screenshots for the specified urls.
 *
 * Usage:
 *  var url = 'http://example.com';
 *  var screenshot = new Screenshot(url);
 *
 *  var urls = ['http://example.com', 'http://example.com/about'];
 *  var screenshots = new Screenshot(urls);
 */

 (function(module){
  'use strict';

  var spawn = require('child_process').spawn;
  var fs = require('fs');
  var async = require('async');
  var glob = require("glob");
  var path = require('path');
  var mkdirp = require('mkdirp');
  var handlebars = require('handlebars');

  require('shelljs/global');

  var childArgs = function(url, outputTmp, index, additional) {
    additional = additional || {};

    var options = [
      '--fullsize',
      '--dir=' + outputTmp,
      '--filename=' + index,
      '--clipwidth=1'
    ];

    if(additional.width) {
      options.push('--width=' + additional.width);
    }

    if(additional.height) {
      options.push('--height=' + additional.height);
    }

    if(additional.agent) {
      options.push('--user-agent="' + additional.agent + '"');
    }

    options.push(url);
    
    return options;
  };

  var Screenshot = function(url, options) {
    this.options = options || {
      output: process.cwd(),
      temp: '/tmp/multishot',
    };

    this.options.template = this.options.template || path.resolve(__dirname, '../template/template.html');
    this.options.styles = this.options.styles || {};

    // Setup default styles
    this.setDefaultStyle('background', '#fff');
    this.setDefaultStyle('bodyFontFamily', '"Lato", sans-serif');
    this.setDefaultStyle('bodyFontWeight', '400');
    this.setDefaultStyle('bodyFontColor', '#7f8c8d');
    this.setDefaultStyle('bodyFontSize', '12px');
    this.setDefaultStyle('bodyLineHeight', '17px');
    this.setDefaultStyle('groupMargin', '40px 0');
    this.setDefaultStyle('screenshotImgBorder', '4px solid #d35400');
    this.setDefaultStyle('h2FontSize', '24px');
    this.setDefaultStyle('h2FontColor', '#2C3E50');
    this.setDefaultStyle('h2FontWeight', '700');
    this.setDefaultStyle('h2Margin', '2px 0');
    this.setDefaultStyle('urlFontStyle', 'italic');

    this.groups = ['Default'];
    this.grouped = {};

    this.template = fs.readFileSync(this.options.template, 'utf8');
  };

  Screenshot.prototype = Object.create(require('events').EventEmitter.prototype);

  Screenshot.prototype.start = function() {
    var self = this;

    if (!which('webkit2png')) {
      this.emit('err', 'This script requires webkit2png. http://www.paulhammond.org/webkit2png/');
      return;
    }

    this.emit('start');

    async.parallel([
      function(done) {
        mkdirp(self.options.output, done);
      },
      function(done) {
        mkdirp(self.options.temp, done);
      },
      function(done) {
        glob(self.options.temp + '/**/*.png', function(err, files) {
          if(err) {
            self.emit('err');
            done();
            return;
          }

          async.each(files, function(file, finish) {
            fs.unlink(file, finish);
          }, function(err) {
            if(err) {
              self.emit('err', err);
            }
            done();
          });
        });
      }
    ],
    function(){
      this.urls = (typeof url === 'string') ? [url] : url;

      this.validateGroups();

      // Process urls
      async.each(this.urls, this.processUrl.bind(this), this.screenshotsComplete.bind(this));
    }.bind(this));
  };

  Screenshot.prototype.validateGroups = function() {
    var i, c, url;

    for(i = 0, c = this.urls.length; i < c; i++) {
      url = this.urls[i];
      if((!!url) && (url.constructor === Object)) {
        if(!url.url) {
          // Falls back to nothing. Will be skipped.
          url.url = '';
        }

        if(!url.group) {
          url.group = 'Default';
        }
      } else {
        url = {
          group: 'Default',
          url: url
        };
      }

      url.index = i;

      this.urls[i] = url;

      if(this.groups.indexOf(url.group) === -1) {
        this.groups.push(url.group);
      }

      if(!this.grouped[url.group]) {
        this.grouped[url.group] = [];
      }

      this.grouped[url.group].push(url);
    }
  };

  Screenshot.prototype.processUrl = function(url, done) {
    // Check if item is empty (either empty param or blank line)
    if(!url) {
      this.emit('progress');
      done();
      return;
    }

    var parser = spawn('webkit2png', childArgs(url.url, path.normalize(this.options.temp + '/' + url.group + '/'), url.index, this.options.additional));
    parser.on('exit', function(code) {
      this.emit('progress');
      done();
    }.bind(this));
  };

  Screenshot.prototype.screenshotsComplete = function() {
    var self = this;
    
    var data = {
      base: this.options.temp,
      grouped: this.grouped,
      styles: this.options.styles
    };

    var template = handlebars.compile(this.template);
    var html = template(data);
    var templateFile = path.normalize(this.options.temp + '/template.html');

    var date = new Date();
    var outputFile = "" + date.getFullYear() + (date.getMonth()+1) + date.getDate();

    fs.writeFileSync(templateFile, html, 'utf8');

    var parser = spawn('webkit2png', childArgs(templateFile, this.options.output, outputFile));
    parser.on('exit', function(code) {
      self.emit('complete', path.normalize(self.options.output + '/' + outputFile + '-full.png'));
    });
  };

  Screenshot.prototype.setDefaultStyle = function(style, value) {
    value = value || null;
    this.options.styles[style] = this.options.styles[style] || value;
  };

  module.exports = Screenshot;
 })(module);